internal/controller/webhook.go

```diff
@@ -0,0 +1,148 @@
+)
+
+const (
+	WebhookServiceName = "odf-operator-webhook-service"
+	WebhookName        = "csv.odf.openshift.io"
+)
+
+func (r *StorageSystemReconciler) reconcileSubscriptionValidatingWebhook() error {
+
+	csvWebhook := admrv1.MutatingWebhook{
+		Name: WebhookName,
+		ClientConfig: admrv1.WebhookClientConfig{
+			Service: &admrv1.ServiceReference{
+				Name: WebhookServiceName,
+				Path: ptr.To(odfwebhook.WebhookPath),
+				Port: ptr.To(int32(443)),
+			},
+		},
+		Rules: []admrv1.RuleWithOperations{
+			{
+				Rule: admrv1.Rule{
+					APIGroups:   []string{"operators.coreos.com"},
+					APIVersions: []string{"v1alpha1"},
+					Resources:   []string{"clusterserviceversions"},
+					Scope:       ptr.To(admrv1.NamespacedScope),
+				},
+				Operations: []admrv1.OperationType{admrv1.Create},
+			},
+		},
+		SideEffects:             ptr.To(admrv1.SideEffectClassNone),
+		TimeoutSeconds:          ptr.To(int32(30)),
+		AdmissionReviewVersions: []string{"v1"},
+		// fail the validation if webhook can't be reached
+		FailurePolicy: ptr.To(admrv1.Fail),
+	}
+
+	whConfig := &admrv1.MutatingWebhookConfiguration{}
+	whConfig.Name = csvWebhook.Name
+
+	res, err := controllerutil.CreateOrUpdate(r.context, r.Client, whConfig, func() error {
+
+		// openshift fills in the ca on finding this annotation
+		whConfig.Annotations = map[string]string{
+			"service.beta.openshift.io/inject-cabundle": "true",
+		}
+
+		var caBundle []byte
+		if len(whConfig.Webhooks) == 0 {
+			whConfig.Webhooks = make([]admrv1.MutatingWebhook, 1)
+		} else {
+			// do not mutate CA bundle that was injected by openshift
+			caBundle = whConfig.Webhooks[0].ClientConfig.CABundle
+		}
+
+		// webhook desired state
+		var wh *admrv1.MutatingWebhook = &whConfig.Webhooks[0]
+		csvWebhook.DeepCopyInto(wh)
+
+		wh.Name = whConfig.Name
+		// only send requests received from own namespace
+		wh.NamespaceSelector = &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				"kubernetes.io/metadata.name": r.OperatorNamespace,
+			},
+		}
+		// preserve the existing (injected) CA bundle if any
+		wh.ClientConfig.CABundle = caBundle
+		// send request to the service running in own namespace
+		wh.ClientConfig.Service.Namespace = r.OperatorNamespace
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	r.logger.Info("successfully created or updated webhook", "operation", res, "name", whConfig.Name)
+	return nil
+}
+
+func (r *StorageSystemReconciler) reconcileWebhookService() error {
+
+	webhookService := corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      WebhookServiceName,
+			Namespace: r.OperatorNamespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Name:       "https",
+					Port:       443,
+					Protocol:   corev1.ProtocolTCP,
+					TargetPort: intstr.FromInt32(9443),
+				},
+			},
+			Selector: map[string]string{
+				"app.kubernetes.io/name": "odf-operator-webhook",
+			},
+			Type: corev1.ServiceTypeClusterIP,
+		},
+	}
```

Comment:
this whole part is basically a template which doesn't need to be instantiated everytime in a function and so pls move this to a templates package and mutate a deepcopy whenever required in createorupdate method.